consulresolver/src/main/java/io/datanerds/grpc/ConsulNameResolver.java

```diff
@@ -0,0 +1,116 @@
+package io.datanerds.grpc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.net.HostAndPort;
+import com.orbitz.consul.Consul;
+import com.orbitz.consul.HealthClient;
+import com.orbitz.consul.cache.ServiceHealthCache;
+import com.orbitz.consul.cache.ServiceHealthKey;
+import com.orbitz.consul.model.ConsulResponse;
+import com.orbitz.consul.model.health.ServiceHealth;
+import com.orbitz.consul.option.ImmutableCatalogOptions;
+import io.grpc.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class ConsulNameResolver extends NameResolver {
+    public static final String DEFAULT_ADDRESS = "localhost";
+    public static final Integer DEFAULT_PORT = 8383;
+    private static final Logger logger = LoggerFactory.getLogger(ConsulNameResolver.class);
+    private final String authority;
+    private final String service;
+    @GuardedBy("this")
+    private Listener listener;
```

Comment:
does not look like listener and healthClient are actually protected / explicitly made thread safe. Does it need to be thread safe?